ibis/expr/datatypes/core.py

```diff
@@ -78,6 +78,198 @@ def castable(self, other):
 
         return castable(self, other)
 
+    @property
+    def is_array(self) -> bool:
+        return isinstance(self, Array)
+
+    @property
+    def is_binary(self) -> bool:
+        return isinstance(self, Binary)
+
+    @property
+    def is_boolean(self) -> bool:
+        return isinstance(self, Boolean)
+
+    @property
+    def is_category(self) -> bool:
+        return isinstance(self, Category)
+
+    @property
+    def is_date(self) -> bool:
+        return isinstance(self, Date)
+
+    @property
+    def is_decimal(self) -> bool:
+        return isinstance(self, Decimal)
+
+    @property
+    def is_enum(self) -> bool:
+        return isinstance(self, Enum)
+
+    @property
+    def is_float16(self) -> bool:
+        return isinstance(self, Float16)
+
+    @property
+    def is_float32(self) -> bool:
+        return isinstance(self, Float32)
+
+    @property
+    def is_float64(self) -> bool:
+        return isinstance(self, Float64)
```

Comment:
Do you have a use case in mind for distinct methods for every datatype? The main use case I have for something like this is detecting classes of datatypes (e.g. `is_numeric()`, `is_floating()`, `is_temporal()`) rather than anything specific like `is_float64()`. Having a method for everything is fine, just curious about the use case.

Reply:
I generated these methods and didn't want to curate them to arbitrarily exclude things.

It also seems a bit arbitrary to include leaf type versions of these for arrays (for example) but then exclude other leaf types.

Strings in theory are a leaf type but right now the JSON type subclasses it. However, if String were a leaf type I think we'd want an `is_string`.

I'll replace the cases of instance checking in the codebase where I can to see which of these are most used. This may inform how valuable (or not) having some of these more specific methods is.